flocker/node/script.py

```diff
@@ -281,143 +320,312 @@ def get_service(self, reactor, options):
 
-    This is a temporary script, until the volume service can be created in
-    ``zfs_dataset_deployer``. The majority of this script will be in
-    ``flocker_dataset_agent_main`` and ``AgentScript``. See FLOC-1924.
-    """
-    def main(self, reactor, options, volume_service):
-        agent_config = options[u'agent-config']
-        configuration = yaml.safe_load(agent_config.getContent())
+    get_external_ip = field(initial=_get_external_ip, mandatory=True)
+
+    control_service_host = field(type=bytes, mandatory=True)
+    control_service_port = field(type=int, mandatory=True)
+
+    # Cannot use type=NodeCredential because one of the tests really wants to
+    # set this to None.
+    node_credential = field(mandatory=True)
+    # Cannot use type=Certificate; pyrsistent rejects classic classes.
+    ca_certificate = field(mandatory=True)
+
+    backend_name = field(type=unicode, mandatory=True)
+    api_args = field(type=PMap, factory=pmap, mandatory=True)
+
+    @classmethod
+    def from_configuration(cls, configuration):
+        """
+        Load configuration from a data structure loaded from the configuration
+        file and only minimally processed.
+
+        :param dict configuration: Agent configuration as returned by
+            ``get_configuration``.
 
-        validate_configuration(configuration=configuration)
+        :return: A new instance of ``cls`` with values loaded from the
+            configuration.
+        """
+        host = configuration['control-service']['hostname']
+        port = configuration['control-service']['port']
+
+        node_credential = configuration['node-credential']
+        ca_certificate = configuration['ca-certificate']
+
+        api_args = configuration['dataset']
+        backend_name = api_args.pop('backend')
 
-        deployer_factory = dataset_deployer_from_configuration(
-            dataset_configuration=configuration['dataset'],
-            volume_service=volume_service
+        return cls(
+            control_service_host=host,
+            control_service_port=port,
+
+            node_credential=node_credential,
+            ca_certificate=ca_certificate,
+
+            backend_name=backend_name.decode("ascii"),
+            api_args=api_args,
         )
 
-        service_factory = AgentServiceFactory(
-            deployer_factory=deployer_factory
-        ).get_service
+    def get_backend(self):
+        """
+        Find the backend in ``self.backends`` that matches the one named by
+        ``self.backend_name``.
 
-        service = service_factory(reactor, options)
+        :raise ValueError: If ``backend_name`` doesn't match any known backend.
+        :return: The matching ``BackendDescription``.
+        """
+        for backend in self.backends:
+            if backend.name == self.backend_name:
+                return backend
+        raise ValueError(
+            "Backend named {} not available".format(self.backend_name),
```

Comment:
:question: Maybe use `{!r}` here to reveal whitespace in the supplied backend name. eg

```
end\n    \"Backend named {!r} not available\".format(self.backend_name),\nexceptions.ValueError: Backend named u'loopback ' not available\n", "message_type": "twisted:log", "task_level": [3]}
{"task_uuid": "a6004bba-8680-49b3-b8e0-163c039f851f", "error": false, "timestamp": 1432642647.106184, "message": "Main loop terminated.", "message_type": "twisted:log", "task_level": [4]}
```

```
(1925)(storage-driver-configuration-FLOC-1925 ✕?)[~/.../HybridLogic/flocker]$ cat agent.yml
control-service:
    hostname: 'localhost'
    port: 4524
dataset:
    backend: "loopback "
version: 1
```